Menus/SpatialUI/Scripts/SpatialUIController.cs

```diff
@@ -0,0 +1,106 @@
+﻿#if UNITY_EDITOR
+using System;
+using System.Reflection;
+using UnityEditor.Experimental.EditorVR.Core;
+using UnityEngine;
+using UnityEngine.InputNew;
+
+namespace UnityEditor.Experimental.EditorVR.Menus
+{
+    /// <summary>
+    /// Mandates that derived classes implement core required SpatialUI controller functionality
+    /// </summary>
+    public abstract class SpatialUIController : MonoBehaviour, INodeToRay
+    {
+        // Pre-box fields to avoid allocation when setting them via reflection
+        static object[] boxedTrueBool = new object[] {true};
+        static object[] boxedFalseBool = new object[] {false};
+        static bool s_Initialized;
+        static bool s_SceneViewGizmosInOriginalState = true;
+        static bool s_SelectionOutlineWasEnabledOnStart;
+        static bool s_SelectionWireframeWasEnabledOnStart;
+        static Type s_SceneViewGizmoAnnotationUtilityType;
+        static Type s_AnnotationUtilityType;
+        static string s_SelectionOutlineProperty = "showSelectionOutline";
+        static string s_SelectionWireframeProperty = "showSelectionWire";
+
+        protected bool sceneViewGizmosVisible
+        {
+            set
+            {
+                if (s_SceneViewGizmosInOriginalState == value)
+                    return;
+
+                s_SceneViewGizmosInOriginalState = value;
+
+                if (s_SceneViewGizmosInOriginalState) // Restore scene view gizmo visibility state
+                    SetSceneViewGizmoStates(s_SelectionOutlineWasEnabledOnStart, s_SelectionWireframeWasEnabledOnStart);
+                else // Hide gizmos if they were visible when startging the EXR session
+                    SetSceneViewGizmoStates();
+            }
+        }
+
+        void Awake()
+        {
+            if (!s_Initialized)
+            {
+                CacheSceneViewGizmoStates();
+                s_Initialized = true;
+            }
+        }
+
+        static void CacheSceneViewGizmoStates()
+        {
+            // Cache the selection-outline & selection-wireframe in the SceneView gizmo states
+            // These are set via checkbox in an Editor Scene/Game view/window/panel
+            s_SceneViewGizmosInOriginalState = true;
+            var asm = Assembly.GetAssembly(typeof(Editor));
+            var annotationUtilityType = "UnityEditor.AnnotationUtility";
```

Comment:
Annotation in this case refers to annotating the data around these controllers, and not the annotation tool, yah?